alchemist/alchemist-incarnation-scafi/src/main/scala/it/unibo/alchemist/model/ScafiIncarnation.scala

```diff
@@ -61,35 +84,66 @@ sealed class ScafiIncarnation[P <: Position[P]] extends Incarnation[Any, P]{
      */
     val doCacheValue = !v.startsWith("_");
-    CachedInterpreter[AnyRef](if(doCacheValue) v else v.tail, doCacheValue)
+    CachedInterpreter[AnyRef](if(doCacheValue) v else v.tail, doCacheValue).asInstanceOf[T]
   }
 
-  override def createCondition(rand: RandomGenerator, env: Environment[Any, P] , node: Node[Any], time: TimeDistribution[Any], reaction: Reaction[Any], param: String) = {
-    throw new UnsupportedOperationException("Use the type/parameters syntax to initialize conditions.")
+  override def createCondition(rand: RandomGenerator, env: Environment[T, P] , node: Node[T], time: TimeDistribution[T], reaction: Reaction[T], param: String): Condition[T] = {
+    if(!node.isInstanceOf[ScafiNode[T,P]]) {
+      throw new IllegalArgumentException(s"The node must be an instance of ${classOf[ScafiNode[_,_]]}"
+      + s", but it is an ${node.getClass} instead.")
+    }
+
+    val alreadyDone = ScafiIncarnationUtils
+      .inboundDependencies(node, classOf[ScafiComputationalRoundComplete[T]])
+      .collect { case x: RunScafiProgram[T,P] => x }
+    val spList = ScafiIncarnationUtils.allScafiProgramsFor(node) -- alreadyDone
+    if (spList.isEmpty) {
+      throw new IllegalStateException("There is no program requiring a " +
+        classOf[ScafiComputationalRoundComplete[_]].getSimpleName + " condition")
+    }
+    if (spList.size > 1) {
+      throw new IllegalStateException("There are too many programs requiring a " +
+        classOf[ScafiComputationalRoundComplete[_]].getName + " condition: " + spList)
+    }
+    new ScafiComputationalRoundComplete(node, spList.head)
   }
 
-  override def createMolecule(s: String ): SimpleMolecule = {
-    new SimpleMolecule(notNull(s))
+  override def createMolecule(s: String): SimpleMolecule = {
+    new SimpleMolecule(notNull(s, "simple molecule name"))
   }
 
-  override def createNode(rand: RandomGenerator, env: Environment[Any, P], param: String) = {
+  override def createNode(rand: RandomGenerator, env: Environment[T, P], param: String) = {
     new ScafiNode(env)
   }
 
-  override def createReaction(rand: RandomGenerator, env: Environment[Any, P], node: Node[Any], time: TimeDistribution[Any], param: String) = {
-    new Event(node, time)
+  override def createReaction(rand: RandomGenerator, env: Environment[T, P], node: Node[T], time: TimeDistribution[T], param: String): Reaction[T] = {
+    import scala.collection.JavaConverters._
+
+    val isSend = "send".equalsIgnoreCase(param)
+    val result: Reaction[T] =
+      if (isSend)
+        new ChemicalReaction[T](Objects.requireNonNull[Node[T]](node), Objects.requireNonNull[TimeDistribution[T]](time))
+      else
+        new Event[T](node, time)
+    if (param != null)
+      result.setActions(ListBuffer[Action[T]](createAction(rand, env, node, time, result, param)).asJava)
+    if (isSend)
+      result.setConditions(ListBuffer[Condition[T]](createCondition(rand, env, node, time, result, null)).asJava)
```

Comment:
Use curly brackets. It's hard to read. Renouncing brackets is ok only in short if expressions.